### PR TITLE
Simplify testing using Ruff

### PR DESCRIPTION
**Pull request by gridhead**
Simplify testing using Ruff

### CODE REVIEW

File: mdapi/database/base.py [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
 
 def index_database(name, tempdtbs):
-    servlogr.logrobjc.info("[%s] Indexing database %s" % (name, tempdtbs))
```
Comment (nphilipp) [inline mdapi/database/base.py:31]: This – https://docs.astral.sh/ruff/rules/logging-percent-format/ – is what ruff flags here, right? In this case (logger objects), the fix is not to use f-strings but retain a percent format string and pass the variables to the method instead of using the percent sign operator:

```diff
-    servlogr.logrobjc.info("[%s] Indexing database %s" % (name, tempdtbs))
+    servlogr.logrobjc.info("[%s] Indexing database %s", name, tempdtbs)
```

This way, `name` and `tempdbs` are only filled into the format string if the log message is used, i.e. if the log level is set accordingly. Because the `logging` module is rather old, it uses percent-style interpolation rather than the new style used by `str.format()` or f-strings.

File: mdapi/database/base.py [hunk 2]
```diff
@@ -80,24 +80,21 @@ def change_row_to_package(self, rowe):
...
                 if rowe[0] in cacheobj:
                     yield (cacheobj[rowe[0]], *rowe[1:])
                 else:
-                    servlogr.logrobjc.debug(
-                        "[%s] %s does not appear in the %s cache for %s"
-                        % (self.name, rowe[0], tableobj, location)
-                    )
-                    servlogr.logrobjc.debug("[%s] Dropping from comparison" % self.name)
+                    servlogr.logrobjc.debug(f"[{self.name}] {rowe[0]} does not appear in the {tableobj} cache for {location}")  # noqa : E501
+                    servlogr.logrobjc.debug(f"[{self.name}] Dropping from comparison")
```
Comment (nphilipp) [inline mdapi/database/base.py:89]: Same thing here and below.

File: mdapi/database/base.py [hunk 3]
```diff
@@ -121,12 +118,10 @@ def main(self):
...
             # so we have nothing to compare it against. Just return and say there
             # are "no differences".
-            servlogr.logrobjc.warning(
-                "[%s] Database empty - %s cannot compare" % (self.name, self.dbsB)
-            )
+            servlogr.logrobjc.warning(f"[{self.name}] Database empty - {self.dbsB} cannot compare")
             return set()
 
-        assert len(tablistA) == len(tablistB), "Cannot compare disparate databases"
+        assert len(tablistA) == len(tablistB), "Cannot compare disparate databases"  # noqa : S101
```
Comment (nphilipp) [inline mdapi/database/base.py]: Here’s documentation about why ruff flags this: https://docs.astral.sh/ruff/rules/assert/

In short, `assert` is a no-op if the Python interpreter runs with the `-O` or `-OO` flags. Here’s an alternative:

```suggestion
        if len(tablistA) != len(tablistB):
            raise ValueError("Cannot compare disparate databases")
```

File: mdapi/database/main.py [hunk 4]
```diff
@@ -43,30 +43,30 @@
...
 def extract_database(name, arcvname, location):
-    servlogr.logrobjc.info("[%s] Extracting %s to %s" % (name, arcvname, location))
+    servlogr.logrobjc.info(f"[{name}] Extracting {arcvname} to {location}")
     if arcvname.endswith(".xz"):
         with lzma.open(arcvname) as inp, open(location, "wb") as otptfile:
             otptfile.write(inp.read())
     elif arcvname.endswith(".tar.gz"):
         with tarfile.open(arcvname) as tararchv:
-            tararchv.extractall(path=location, members=check_archive_validity())
+            tararchv.extractall(path=location, members=check_archive_validity())  # noqa : S202
```
Comment (nphilipp) [inline mdapi/database/main.py]: Is requiring Python >= 3.12 an option? In this case, rather use this:

```suggestion
            tararchv.extractall(path=location, filter="data")
```

Your “workaround” won’t work BTW. 🤷

File: pyproject.toml [hunk 5]
```diff
@@ -45,27 +45,25 @@ pyzstd = "^0.15.0"
...
-black = "^23.0.0 || ^24.0.0"
-isort = "^5.10.1"
+tox = "^4.0.0"
 pytest = "^7.1.3 || ^8.0.0"
-flake8 = "<7.0.1"
-pytest-black = "^0.3.12"
-pytest-flake8 = "^1.0.7"
-pytest-isort = "^3.0.0 || ^4.0.0"
+pytest-cov = "^4.1.0"
+ruff = "^0.0.285 || ^0.1.0 || ^0.2.0 || ^0.3.0"
```
Comment (nphilipp) [inline pyproject.toml]: You use `[tool.ruff.lint]` below. As I understand it, this needs `ruff = "^0.2.0 || ^0.3.0"` here (the old versions expected it all in `[tool.ruff]`)

File: pyproject.toml [hunk 6]
```diff
@@ -45,27 +45,25 @@ pyzstd = "^0.15.0"
...
 addopts = "--black --isort --flake8 --cov=mdapi --cov-report=term --cov-report=xml --cov-report=html"
 flake8-max-line-length = 100
 asyncio_mode = "auto"
 
-[tool.isort]
-line_length = 100
-profile = "black"
+[tool.ruff]
+line-length = 100
+fix = true
```
Comment (nphilipp) [inline pyproject.toml]: Hmm, I wouldn’t do it. A developer could run ruff which changes files and then the changes could slip in unreviewed with something unrelated (e.g. with `git commit -a`).

File: tests/test_clim.py [hunk 7]
```diff
@@ -32,30 +32,28 @@
 def test_cli_application_help_option():
     rnnrobjc = CliRunner()
     rsltobjc = rnnrobjc.invoke(main, ["--help"])
-    assert "A simple API for serving the metadata from the RPM repositories" in rsltobjc.output
-    assert rsltobjc.exit_code == 0
+    assert "A simple API for serving the metadata from the RPM repositories" in rsltobjc.output  # noqa : S101
+    assert rsltobjc.exit_code == 0  # noqa : S101
```
Comment (nphilipp) [inline tests/test_clim.py:38]: You can tell ruff to ignore asserts in tests wholesale by putting this into pyproject.toml:

```
[tool.ruff.lint.per-file-ignores]
"tests/*" = ["S101"]
```

File: tests/__init__.py [hunk 8]
```diff
@@ -30,12 +30,11 @@
 Standard set of databases to run tests against
 """
 
-LOCATION = "/var/tmp/mdapi-tests/"
+LOCATION = "/var/tmp/mdapi-tests/"  # noqa : S108
```
Comment (nphilipp) [inline tests/__init__.py]: Ruff flags this for a good reason 😉 – read e.g. this for some background: https://systemd.io/TEMPORARY_DIRECTORIES/

I’ve skimmed the tests using this and have some ideas how this could be done differently.

File: tests/__init__.py [hunk 9]
```diff
@@ -30,12 +30,11 @@
...
 BRCHNAME = "rawhide"
 PROBEURL = {
-    "koji": "https://kojipkgs.fedoraproject.org/repos/%s/latest/x86_64/repodata/" % BRCHNAME,
-    "rawhide": "https://dl.fedoraproject.org/pub/fedora/linux/development/%s/Everything/x86_64/os/repodata/"  # noqa
-    % BRCHNAME,
+    "koji": f"https://kojipkgs.fedoraproject.org/repos/{BRCHNAME}/latest/x86_64/repodata/",
+    "rawhide": f"https://dl.fedoraproject.org/pub/fedora/linux/development/{BRCHNAME}/Everything/x86_64/os/repodata/",
```
Comment (nphilipp) [inline tests/__init__.py:42]: As discussed, this can break.

File: tox.ini [hunk 10]
```diff
@@ -1,6 +1,6 @@
 [tox]
 minversion = 3.8.0
-envlist = py{38,39,310},bandit,black,flake8,isort
+envlist = py{38,39,310,311,312},bandit,cleaning
```
Comment (nphilipp) [inline tox.ini]: Doesn’t ruff incorporate bandit tests, too?

**Review by nphilipp — CHANGES_REQUESTED** [review]

File: mdapi/database/base.py (hunk 1 above)
Comment (gridhead) [reply mdapi/database/base.py:31]: I wonder if the f-string changes can stay and work just fine with the logging implementation that I have in the project or if I should address the change according to the recommendation you made here. 🤔

File: mdapi/database/base.py (hunk 3 above)
Comment (gridhead) [reply mdapi/database/base.py]: The likeliness of someone running tests in the production environment (i.e. using the `-0` and the `-00` as mentioned above) is small but admittedly, not zero. I wonder if we could let things be the way they are as `S101` does feel a bit nitpicky[1] to me.

Sauce
1. https://stackoverflow.com/a/68429294

File: mdapi/database/main.py (hunk 4 above)
Comment (gridhead) [reply mdapi/database/main.py]: I am reminded of @dralley's [comment](https://github.com/fedora-infra/mdapi/issues/161#issuecomment-1522445080) from some eons ago and am tempted to purge those lines to see if that forsakes the codebase. If it does not, I will make the changes that you mentioned - barring to see if the `filter` option is available in Python versions lower than 3.12.

The workaround does work though. As long as I keep ignoring the said problem, the problem should not affect me. Here is a reference image of me doing just fine with the codebase.

image

File: pyproject.toml (hunk 5 above)
Comment (gridhead) [reply pyproject.toml]: Gotcha.

I will restrict the version requirements to `ruff = "^0.2.0 || ^0.3.0"` and regenerate the lockfile.

File: pyproject.toml (hunk 6 above)
Comment (gridhead) [reply pyproject.toml]: So `fix = false`?

On it.

File: tests/test_clim.py (hunk 7 above)
Comment (gridhead) [reply tests/test_clim.py:38]: TIL we could do that rather than sweating my way across the codebase to see wherever `assert` is used.

File: tests/__init__.py (hunk 8 above)
Comment (gridhead) [reply tests/__init__.py]: I will attempt adding my own implementation using the [`tempfile` library](https://docs.python.org/3/library/tempfile.html) so that you have something to cringe on early in the morning over a mug of coffee.

File: tox.ini (hunk 10 above)
Comment (gridhead) [reply tox.ini]: Does it? I am not too sure.

File: tests/__init__.py (hunk 9 above)
Comment (gridhead) [reply tests/__init__.py:42]: Not sure if you mean that the "line" can break (vide `E501`) or if the downloading part, the failure of which can break the tests? Assuming it is the latter, that is a scenario I am totally fine with as even the downloading part needs testing here.

File: mdapi/database/base.py (hunk 1 above)
Comment (nphilipp) [reply mdapi/database/base.py:31]: The reason why I would not use an f-string here – and I would use them just about everywhere else – is not that it won’t work fine (it will), but that it would always expand the placeholders in the format string, even if logging configuration makes the message not end up anywhere.

Letting the logger (or probably rather the formatter) do the work lazily, only if necessary, is a performance optimization. It’s admittedly awkward that it makes us use two different ways of string interpolation (new-style in f-strings, %-style in log messages), but I’d do it anyway – once it becomes a habit, it comes at no cost.

File: mdapi/database/base.py (hunk 3 above)
Comment (nphilipp) [reply mdapi/database/base.py]: It would be nit-picky if the argument was only about the likelihood of the code being run optimized or not. In my opinion, this is not the case: asserts (in Python and other languages) express that a certain condition applies – if it doesn’t, it must be a bug in the program.

I’m not very familiar with mdapi code, but it looks to me as if it pulls a database from a remote repository, then attempts to recreate it from other metadata and compares the two? In this case, a simple syncing problem on the other end could cause issues here.

File: mdapi/database/main.py (hunk 4 above)
Comment (nphilipp) [reply mdapi/database/main.py]: > I am reminded of @dralley's https://github.com/fedora-infra/mdapi/issues/161#issuecomment-1522445080 from some eons ago and am tempted to purge those lines to see if that forsakes the codebase.

Sounds reasonable.

> barring to see if the filter option is available in Python versions lower than 3.12.

I just checked, and it has been introduced in later releases of the previous minor versions – 3.8.18, 3.9.17, 3.10.13, 3.11.5 should have it.

> The workaround does work though.

Does it? The `check_archive_validity()` function returns `True` (a boolean), `.extractall(members=…)` expects a list of archive members (i.e. files) to be extracted from the tarball.

File: pyproject.toml (hunk 6 above)
Comment (nphilipp) [reply pyproject.toml]: Just drop the line (`fix = false` is the default).

File: tests/__init__.py (hunk 8 above)
Comment (nphilipp) [reply tests/__init__.py]: 😂

File: tests/__init__.py (hunk 9 above)
Comment (nphilipp) [reply tests/__init__.py:42]: The downloading part. Other projects use [custom pytest markers](https://docs.pytest.org/en/latest/example/markers.html#mark-examples) to “tag” which tests can be run offline vs. those that need access to the internet – then the tests to be run can be filtered accordingly.

File: tox.ini (hunk 10 above)
Comment (nphilipp) [reply tox.ini]: Some of the above warnings are from [ruff’s flake8-bandit emulation](https://docs.astral.sh/ruff/rules/#flake8-bandit-s).

File: mdapi/database/main.py [hunk 11]
```diff
@@ -32,65 +32,61 @@
...
 def extract_database(name, arcvname, location):
-    servlogr.logrobjc.info("[%s] Extracting %s to %s" % (name, arcvname, location))
+    servlogr.logrobjc.info(f"[{name}] Extracting {arcvname} to {location}")
     if arcvname.endswith(".xz"):
         with lzma.open(arcvname) as inp, open(location, "wb") as otptfile:
             otptfile.write(inp.read())
     elif arcvname.endswith(".tar.gz"):
         with tarfile.open(arcvname) as tararchv:
-            tararchv.extractall(path=location, members=check_archive_validity())
+            tararchv.extractall(path=location, filter="data")  # noqa : S202
```
Comment (nphilipp) [inline mdapi/database/main.py]: AIUI, with `filter` set, this doesn’t need `noqa` anymore: https://docs.astral.sh/ruff/rules/tarfile-unsafe-members/

File: pyproject.toml [hunk 12]
```diff
@@ -45,27 +45,25 @@ pyzstd = "^0.15.0"
...
 addopts = "--black --isort --flake8 --cov=mdapi --cov-report=term --cov-report=xml --cov-report=html"
 flake8-max-line-length = 100
 asyncio_mode = "auto"
 
-[tool.isort]
-line_length = 100
-profile = "black"
+[tool.ruff]
+line-length = 100
+fix = false
```
Comment (nphilipp) [inline pyproject.toml]: As I mentioned, `fix = false` is the default, so you don’t have to put it in explicitly.

**Review by nphilipp — CHANGES_REQUESTED** [review]

File: mdapi/database/base.py (hunk 3 above)
Comment (gridhead) [reply mdapi/database/base.py]: If only I knew how to read, I would have gotten to know that you marked out the `assert` in a file named `mdapi/database/base.py` and not from a test file. The points I referenced still stand for a testing environments and I sense that we talked in parallels in the aforementioned comments - where I was going about tests and you were about the actual codebase. 

TLDR - Changing it now.

File: mdapi/database/main.py (hunk 4 above)
Comment (gridhead) [reply mdapi/database/main.py]: @nphilipp the last part was a (poorly executed) sarcasm. Check the image of the image of the Ostrich for reference.

File: mdapi/database/base.py (hunk 3 above)
Comment (nphilipp) [reply mdapi/database/base.py]: Yeah, asserts in the test suite are fine!

File: mdapi/database/main.py (hunk 4 above)
Comment (nphilipp) [reply mdapi/database/main.py]: Ah okay 😂 – sometimes I’m slow on the uptake 😉.

**Review by nphilipp — APPROVED** [review]
Looks good to me!
